gradio/queueing.py

```diff
@@ -228,6 +228,9 @@ async def clean_event(self, event: Event) -> None:
             async with self.delete_lock:
                 self.event_queue.remove(event)
 
+        if self.server_app:
+            self.server_app.state_holder.del_if_exists(event.session_hash)
```

Comment:
We're using a threading.lock within a co-routine. Won't this block the event loop?

Reply:
Thank you for pointing out a good point!

I think this won't be a problem because;
* The HTTP server runs in [a single thread](https://github.com/gradio-app/gradio/blob/c9af4f794060e218193935d7213f0991a374f502/gradio/networking.py#L57), so in practice this lock is no-op anytime.
* Even in a multi-threaded case, the code inside the lock is executed very fast, so I think there is no additional cost here.

I would like to ask @aliabid94 for an advice on this, who added `StateHolder` in #5240. Thanks!

Reply:
It probably would be ok in practice but I think we should be using an `asyncio.Lock` instead of a `threading.Lock` in the StateHolder for the sake of adhering to best practices.

The lock in StateHolder is only called within [restore_session_state](https://github.com/gradio-app/gradio/blob/ef96d351229272738fc3c9680f7111f159590341/gradio/route_utils.py#L167), which is called within an [async](https://github.com/gradio-app/gradio/blob/ef96d351229272738fc3c9680f7111f159590341/gradio/route_utils.py#L195) function.

So we can make `restore_session_state` async and use an asyncio.Lock in StateHolder.

Reply:
Yeah it makes sense and I agree that it works perfectly with the current setup where the server dispatches only a single thread.
In turn, if we use `asyncio.Lock` in `.del_if_exists()` here, I think we should remove the lock in `.update()` or replace it with `asyncio.Lock` as well for consistency:
https://github.com/gradio-app/gradio/blob/d6295df030b21b863237e93384b85e19e9c30c30/gradio/state_holder.py#L32
This lock is `threading.Lock` and it implies the `StateHolder` object can be accessed from multiple threads in the future. `asyncio.Lock()` is not thread-safe.

Reply:
Yeah it makes sense and I agree that it works perfectly with the current setup where the server dispatches only a single thread.
In turn, if we change the lock to `asyncio.Lock`, I just have a concern about the original motivation of `threading.Lock()` usage in `StateHolder. It implies multi-thread access in the future? [`asyncio.Lock`](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Lock) is not thread-safe.
It would be nice if we can ask @aliabid94 about it.

Reply:
@aliabid94 was OOO last week and is OOO for a few more days, if this is blocking you @whitphx I think it should be fine to switch to asyncio.Lock and we can ping @aliabid94 when he's back to confirm.

Reply:
Or, if we stop using `threading.Lock` which means we assume the code is not multi-threaded, we can omit the lock itself rather than replacing it to `asyncio.Lock`.
The locked part of `.update()` and `.del_if_exists()` only include synchronous code, so due to the non-preemptive nature of asyncio event loops, other coroutines never interrupt their executions in a single-threaded scenario.

Reply:
In turn, if we decide to use `threading.Lock`, I think it's still ok because there is no additional cost compared to `asyncio.Lock()` in this case because the locked part only contains synchronous and atomic operations which never yield even in async code anyway (I described this as just "very fast" in the first comment which was not precise, sorry).

Reply:
@whitphx Thanks for the detailed explanation. I'm convinced we can remove the lock, let's use https://github.com/gradio-app/gradio/commit/00e39daaeaa94732bb290388fe48071d801788e0